File: 6. Zigzag Conversion/6. Zigzag Conversion.rb
def convert(s, num_rows)
    return s if num_rows == 1
    
    rows = Array.new([num_rows, s.length].min) { '' }
    cur_row = 0
    going_down = false
    
    s.each_char do |c|
        rows[cur_row] += c
        if cur_row == 0 || cur_row == num_rows - 1
            going_down = !going_down
        end
        cur_row += going_down ? 1 : -1
    end
    
    rows.join
end

File: 6. Zigzag Conversion/6. Zigzag Conversion.cs
public class Solution {
    public string Convert(string s, int numRows) {
        if (numRows == 1) return s;
        
        var rows = new List<StringBuilder>();
        for (int i = 0; i < Math.Min(numRows, s.Length); i++)
            rows.Add(new StringBuilder());
        
        int curRow = 0;
        bool goingDown = false;
        
        foreach (char c in s) {
            rows[curRow].Append(c);
            if (curRow == 0 || curRow == numRows - 1) goingDown = !goingDown;
            curRow += goingDown ? 1 : -1;
        }
        
        var result = new StringBuilder();
        foreach (var row in rows)
            result.Append(row);
        
        return result.ToString();
    }
}